### ISSUE TITLE
`string` reading from <&0 in a piped function always reads from the tty

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#5714]
I'm not sure if we're already tracking this as one of the other fd-related issues, but attempting to read from `&0` in a function is interpreted as `/dev/tty` and not the piped input to the function.

```fish
function foo
    string match -r 'world'  <&0
end

echo "hello world" | foo
```

Under fish 2.7.1, this hangs indefinitely (regardless of ^C, ^D, or ^Z) . Under fish 3.0 with the job management fixes, it connects to the tty and takes the input from there (until ^D).

Because of #5457, it's required to explicitly pipe input into `string`. But attempting to do so via handle manipulation just breaks everything.

The workaround is to "abuse" `cat`, which reads from the correct `stdin`:

```fish
function foo
    cat - | string match -r 'world'  <&0
end

echo "hello world" | foo
```

This isn't an issue with fds in general, because this extremely pointless example works fine:

```fish
function foo
    cat - | cat - <&0
end

echo "hello world" | foo
```

(`read -z` may be substituted for `cat -` in any of the examples above)

**Comment by faho** [comment]
Sooo... `<&0` looks to me like it should be a noop, like `>&1` and `2>&2`. 

"Redirect stdin to wherever fd 0 is pointing to", is, given that stdin is fd 0, entirely redundant.

I'd wager we still set the "directly redirected" flag, though, because it's a redirection, and then we miss some redirection (see e.g #206).

Instead of requiring this weird bit of syntax (which could even be a syntax error as it's entirely redundant), why not add a `--stdin` flag to string? (I wouldn't use `-` as an indicator because `string` doesn't take files as arguments)

**Comment by mqudsi** [comment]
> Sooo... `<&0` looks to me like it should be a noop, like `>&1` and `2>&2`.

Yes, it *should* be. I'm pretty sure that is not strictly the case when dealing with blocks; I don't have any concrete examples on me right now, but I'm pretty sure in the past I've had to use it within a block to get a command to read from the block's stdin rather than the terminal (but that behavior may have changed with the improvements to the job control), which is the only reason I tried it in the first place.

I'm not in favor of expanding the api by add flags to support what should be simple "if no input is provided, use stdin" behavior. This isn't restricted to `string`, the concerns about null concatenation entirely voiding the presence of non-quoted strings applies elsewhere (to fish builtins and to external commands) and I don't think we should be going to such lengths to prevent it from reading from stdin.

**Comment by faho** [comment]
>support what should be simple "if no input is provided, use stdin" behavior

That would mean that `string something -- $var` would suddenly read from stdin if $var happens to be empty.

Which is *not* the same as "null concatenation" (aka "cartesian product") - empty variables alone being removed is the same in posix shells, and completely uncontroversial. (same for command substitutions with empty output)

And now `string` would suddenly at least start reading stdin, or at worst read the terminal - which means it would appear to _hang_!

Compare `source`, where we now require a literal `-` argument to read from the terminal, because people used `source (somecommand)` in their config.fish, and that hung.

So I'm strongly in favor of a nice and explicit flag.

Unless you can propose some behavior to reliably detect when reading from stdin is okay? Reading from the tty when you don't want it is particularly bad (because there's no indication of it happening and it needs user interaction), so that's right out, but are there other kinds of stdin that you do or don't want to read?

I can see how in `something | begin; string something; end` you'd want it to read from stdin, but what makes that any different from `something | begin; string something -- $var | somethingelse; read somevar; end` where you want stdin to go to `read`? And especially, what separates them from inside string, where `$var` has already entirely disappeared? And how do you make a nice, concise and understandable rule for when and how it reads from stdin?

**Comment by mqudsi** [comment]
> Which is not the same as "null concatenation" (aka "cartesian product") - empty variables alone being removed is the same in posix shells, and completely uncontroversial. (same for command substitutions with empty output)

I think there is an argument to be made for breaking with posix in this regard, but then again, I'm also sure I'm glossing over some very useful cases of that behavior that we rely on regularly.

**Comment by mattmc3** [comment]
Per #10421, this flag would not just be useful for `string`, but `math` as well.

**Comment by tmccombs** [comment]
Not sure if it's a good idea but another option could be to assume that if "--" is passed in, the user doesn't want it to read from stdin, and provided a possibly empty argument
